### DISCUSSION TITLE
How to import OpaqueToken type?

### DISCUSSION THREAD

**Discussion by D-Marc1** [adonisjs/core#2444] [Help]
Hi, very simple question, as the title states. Not sure if it's obvious, but how would one import `OpaqueToken` type? I don't see it anywhere in the docs, but I do see a link to the class `OpaqueToken` on GitHub. I'm just not sure how I would import it.

**Comment by thetutlage** [comment]
You can please share the use case or what exactly you are trying to achieve?

**Reply by D-Marc1** [reply]
So say I'm authenticating with the API guard.

I would do something like:

```
import { OpaqueTokenContract } from '@ioc:Adonis/Addons/Auth'

const token: OpaqueTokenContract<User> = await auth.attempt(email, password)

const tokenJSON = token.toJSON()
```

I'd like to be able to use `OpaqueToken`, like so: `const tokenJSON: OpaqueToken = token.toJSON()`. My question is, how can I import this type?

**Reply by thetutlage** [reply]
What you have written should work.

Also, I am not sure why you really have to self define the type, coz if you use the guard explicitly, it will infer the type for you.

**Reply by D-Marc1** [reply]
But say I need to add the type on the return? My linter is saying I need to add a type to my return.

so say I have the following method:

```ts
public async login({ request }: HttpContextContract) {
  const { email, password } = request.only(['email', 'password'])

  const token: OpaqueTokenContract<User> = await auth.attempt(email, password)

  const tokenJSON = token.toJSON()

  return tokenJSON
}
```

I'm getting:

> Missing return type on function.eslint@typescript-eslint/explicit-module-boundary-types

**Reply by thetutlage** [reply]
Its because you are using a rule that forces you to define the return type of all the methods. https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-module-boundary-types.md 

Using such a rule will expect you to manually add types everywhere, even when the code can infer it. Again, its a personal choice, and if you stick with the lint rule, then yes you need manual types eveywhere

**Reply by D-Marc1** [reply]
Ya, I'm still debating how strict I want things. But back to the question, is there any way to import `OpaqueToken` type?

**Reply by thetutlage** [reply]
Shared here that your code should work.

https://github.com/adonisjs/core/discussions/2444#discussioncomment-573076

**Reply by D-Marc1** [reply]
I know it works. Was just wondering if it's possible to import `OpaqueToken` out of curiosity.
